storage/innobase/log/log0log.cc

```diff
@@ -1002,13 +1012,14 @@ template<bool release_latch> inline lsn_t log_t::write_buf() noexcept
                         buf + length, flush_buf);
       ... /* TODO: Update the LSN and adjust other code. */
 #else
-# ifdef HAVE_valgrind
       MEM_MAKE_DEFINED(buf + length, (write_size_1 + 1) - length);
-      if (UNIV_LIKELY_NULL(re_write_buf))
-        MEM_MAKE_DEFINED(re_write_buf + length, (write_size_1 + 1) - length);
-# endif
-      buf[length]= 0; /* allow recovery to catch EOF faster */
+      buf[length]= 0; /* ensure that recovery catches EOF */
 #endif
+      if (resizing < 0 ? UNIV_UNLIKELY(!!re_write_buf) : bool(resizing))
+      {
+        MEM_MAKE_DEFINED(re_write_buf + length, (write_size_1 + 1) - length);
+        re_write_buf[length]= 0;
```

Comment:
For the record: I tested commenting out both `re_write_buf[…]= 0` in this function, and then running the test `innodb.log_file_size_online`. The test passed 140×10 repetitions on a powerful server. I double checked that this interface was being used, instead of `cmake -DWITH_INNODB_PMEM=ON`. Finally, on another attempt (140×50 tests), I got a failure (not the one that I was looking for):
```
innodb.log_file_size_online 'encrypted'  w127 [ 8 fail ]
        Test ended at 2024-12-10 04:13:03

CURRENT_TEST: innodb.log_file_size_online
mysqltest: At line 51: query 'reap' failed: <Unknown> (2013): Lost connection to server during query
…
#6  0x00005bb38395248c in memcpy (__len=0x1c0, __src=0x200, __dest=<optimized out>) at /usr/include/x86_64-linux-gnu/bits/string_fortified.h:29
#7  memcpy_aligned<16ul> (n=0x1c0, src=0x200, dest=<optimized out>) at /home/marko/10.6/include/assume_aligned.h:44
#8  log_t::write_buf<1> (this=0x5bb384ca2080 <log_sys>) at /home/marko/10.6/storage/innobase/log/log0log.cc:1045
#9  log_writer<true> (lsn=<optimized out>, durable=0x1, callback=<optimized out>) at /home/marko/10.6/storage/innobase/log/log0log.cc:1166
#10 0x00005bb3839e5f8f in log_t::write_up_to (callback=0x7436697673c0, durable=<optimized out>, lsn=0x3680df5, this=<optimized out>) at /home/marko/10.6/storage/innobase/include/log0log.h:569
#11 trx_flush_log_if_needed (lsn=0x3680df5, trx=trx@entry=0x74366ac01680) at /home/marko/10.6/storage/innobase/trx/trx0trx.cc:1275
#12 0x00005bb3839e7196 in trx_flush_log_if_needed (trx=0x74366ac01680, lsn=<optimized out>) at /home/marko/10.6/storage/innobase/trx/trx0trx.cc:1262
#13 0x00005bb383908ce3 in innobase_commit (hton=<optimized out>, thd=0x7435ec000c68, commit_trx=0x0) at /home/marko/10.6/storage/innobase/handler/ha_innodb.cc:4570
#14 0x00005bb3836cb200 in commit_one_phase_2 (thd=thd@entry=0x7435ec000c68, all=0x0, trans=0x7435ec004518, is_real_trans=0x1) at /home/marko/10.6/sql/handler.cc:2154
#15 0x00005bb3836cb3a9 in ha_commit_one_phase (thd=thd@entry=0x7435ec000c68, all=all@entry=0x0) at /home/marko/10.6/sql/handler.cc:2107
#16 0x00005bb3836cc347 in ha_commit_trans (thd=thd@entry=0x7435ec000c68, all=all@entry=0x0) at /home/marko/10.6/sql/handler.cc:1901
#17 0x00005bb3835b8385 in trans_commit_stmt (thd=thd@entry=0x7435ec000c68) at /home/marko/10.6/sql/transaction.cc:497
#18 0x00005bb38348d93c in mysql_execute_command (thd=thd@entry=0x7435ec000c68, is_called_from_prepared_stmt=is_called_from_prepared_stmt@entry=0x0) at /home/marko/10.6/sql/sql_parse.cc:6216
#19 0x00005bb38349287b in mysql_parse (thd=0x7435ec000c68, rawbuf=<optimized out>, length=<optimized out>, parser_state=<optimized out>) at /home/marko/10.6/sql/sql_parse.cc:8179
#20 0x00005bb38349488a in dispatch_command (command=command@entry=COM_QUERY, thd=thd@entry=0x7435ec000c68, packet=packet@entry=0x7435ec008579 "UPDATE t SET b='' WHERE a<10", packet_length=packet_length@entry=0x1c, blocking=blocking@entry=0x1) at /home/marko/10.6/sql/sql_parse.cc:2006
```
We have `log_sys.resize_flush_buf=nullptr` and `re_write_buf=nullptr` here, as well as `log_sys.resize_lsn=0`. Apparently, log resizing had been completed.

I realize that this version of the code is dereferencing `log_t::writer` while not holding `log_sys.latch`. It looks like it would only be safe to put the following part of `log_write_up_to()` behind a function pointer:
```c++
pending_write_lsn= write_lock.release(log_sys.write_buf<resizing>());
```

Reply:
After I made `log_sys.writer` wrap instantiations of the function `log_sys.write_buf`, the test `innodb.log_file_size_onilne` passed 140×50 rounds both with and without the following patch:
```diff
diff --git a/storage/innobase/log/log0log.cc b/storage/innobase/log/log0log.cc
index 9a851c41e67..a0b99211597 100644
--- a/storage/innobase/log/log0log.cc
+++ b/storage/innobase/log/log0log.cc
@@ -1018,7 +1018,6 @@ lsn_t log_t::write_buf() noexcept
       if (resizing < 0 ? UNIV_UNLIKELY(!!re_write_buf) : bool(resizing))
       {
         MEM_MAKE_DEFINED(re_write_buf + length, (write_size_1 + 1) - length);
-        re_write_buf[length]= 0;
       }
       length= write_size_1 + 1;
     }
@@ -1044,7 +1043,6 @@ lsn_t log_t::write_buf() noexcept
                            new_buf_free);
           memcpy_aligned<16>(resize_flush_buf, re_write_buf + length,
                              (new_buf_free + 15) & ~15);
-          re_write_buf[length + new_buf_free]= 0;
         }
         length+= write_size_1 + 1;
       }
```
I tested this patch also on my local system. In both cases, I failed to reproduce the problem that this patch aims to fix. @Thirunarayanan can you please test my fix with and without the above?

Reply:
yes, After applying the above patch(removing 0 in re_write_buf), I can repeat the scenario.